### PR TITLE
cache cloud id data in CloudIdManager

### PR DESCRIPTION
**Pull request by icewind1991**
Instead of either getting the local display name or searching for a calendar entry every time, cache things for 15m.
Cache is invalidated when a user changes his displayname or the calendar entry gets changed

### CODE REVIEW

**Review by PVince81 — APPROVED** [review]
:+1: ok in general, nice to see that many tests

see 2 minor comments

File: lib/private/Federation/CloudIdManager.php [hunk 1]
```diff
@@ -43,11 +49,48 @@ class CloudIdManager implements ICloudIdManager {
...
+	private ICache $memCache;
+	/** @var array[] */
+	private array $cache = [];
+
+	public function __construct(
+		IManager $contactsManager,
+		IURLGenerator $urlGenerator,
+		IUserManager $userManager,
+		ICacheFactory $cacheFactory,
+		EventDispatcherInterface $eventDispatcher
```
Comment (ChristophWurst) [inline lib/private/Federation/CloudIdManager.php:61]: It should be possible to use our abstraction ``\OCP\EventDispatcher\IEventDispatcher``.

File: lib/private/Federation/CloudIdManager.php [hunk 2]
```diff
@@ -43,11 +49,48 @@ class CloudIdManager implements ICloudIdManager {
...
+	) {
 		$this->contactsManager = $contactsManager;
 		$this->urlGenerator = $urlGenerator;
 		$this->userManager = $userManager;
+		$this->memCache = $cacheFactory->createDistributed('cloud_id_');
```
[the remark this answers is not in the document]
Comment (icewind1991) [reply lib/private/Federation/CloudIdManager.php:66]: It's distributed to make the cache invalidation that happen on displayname/adressbook change propagate.

The distributed+local caching is also used in the `DisplayNameCache`

File: lib/private/Federation/CloudIdManager.php [hunk 3]
```diff
@@ -120,18 +163,42 @@ protected function getDisplayNameFromContact(string $cloudId): ?string {
...
+			// note that for remote id's we don't strip the protocol for the remote we use to construct the CloudId
+			// this way if a user has an explicit non-https cloud id this will be preserved
+			// we do still use the version without protocol for looking up the display name
 			$fixedRemote = $this->fixRemoteURL($remote);
 			$host = $this->removeProtocolFromUrl($fixedRemote);
+		}
+
+		$key = $user . '@' . ($isLocal ? 'local' : $host);
+		$cached = $this->cache[$key] ?? $this->memCache->get($key);
+		if ($cached) {
```
[the remark this answers is not in the document]
Comment (icewind1991) [reply lib/private/Federation/CloudIdManager.php:181]: The cached value is an array, so "falsy" values are not a problem

File: lib/private/Federation/CloudIdManager.php (hunk 1 above)
Comment (icewind1991) [reply lib/private/Federation/CloudIdManager.php]: done

File: apps/files_sharing/tests/External/ManagerTest.php [hunk 4]
```diff
@@ -54,6 +55,7 @@
 use OCP\IUserSession;
 use OCP\Share\IShare;
 use Psr\Log\LoggerInterface;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
```
Comment (CarlSchwan) [inline apps/files_sharing/tests/External/ManagerTest.php:58]:
```suggestion
 use OCP\EventDispatcher\IEventDispatcher;
```

File: apps/files_sharing/tests/External/ManagerTest.php [hunk 5]
```diff
@@ -130,7 +132,13 @@ protected function setUp(): void {
 
 		$this->testMountProvider = new MountProvider(\OC::$server->getDatabaseConnection(), function () {
 			return $this->manager;
-		}, new CloudIdManager($this->contactsManager, $this->createMock(IURLGenerator::class), $this->userManager));
+		}, new CloudIdManager(
+			$this->contactsManager,
+			$this->createMock(IURLGenerator::class),
+			$this->userManager,
+			$this->createMock(ICacheFactory::class),
+			$this->createMock(EventDispatcherInterface::class)
```
Comment (CarlSchwan) [inline apps/files_sharing/tests/External/ManagerTest.php:140]:
```suggestion
			$this->createMock(IEventDispatcher::class)
```

File: apps/files_sharing/tests/External/CacheTest.php [hunk 6]
```diff
@@ -68,7 +70,13 @@ protected function setUp(): void {
 
 		$this->contactsManager = $this->createMock(IManager::class);
 
-		$this->cloudIdManager = new CloudIdManager($this->contactsManager, $this->createMock(IURLGenerator::class), $this->createMock(IUserManager::class));
+		$this->cloudIdManager = new CloudIdManager(
+			$this->contactsManager,
+			$this->createMock(IURLGenerator::class),
+			$this->createMock(IUserManager::class),
+			$this->createMock(ICacheFactory::class),
+			$this->createMock(EventDispatcherInterface::class)
```
Comment (CarlSchwan) [inline apps/files_sharing/tests/External/CacheTest.php:78]:
```suggestion
			$this->createMock(IEventDispatcher::class)
```

File: apps/files_sharing/tests/External/CacheTest.php [hunk 7]
```diff
@@ -31,8 +31,10 @@
 use OCA\Files_Sharing\Tests\TestCase;
 use OCP\Contacts\IManager;
 use OCP\Federation\ICloudIdManager;
+use OCP\ICacheFactory;
 use OCP\IURLGenerator;
 use OCP\IUserManager;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
```
Comment (CarlSchwan) [inline apps/files_sharing/tests/External/CacheTest.php:37]:
```suggestion
 use OCP\EventDispatcher\IEventDispatcher;
```

File: apps/federatedfilesharing/tests/FederatedShareProviderTest.php [hunk 8]
```diff
@@ -116,7 +118,13 @@ protected function setUp(): void {
 		//$this->addressHandler = new AddressHandler(\OC::$server->getURLGenerator(), $this->l);
 		$this->addressHandler = $this->getMockBuilder('OCA\FederatedFileSharing\AddressHandler')->disableOriginalConstructor()->getMock();
 		$this->contactsManager = $this->createMock(IContactsManager::class);
-		$this->cloudIdManager = new CloudIdManager($this->contactsManager, $this->createMock(IURLGenerator::class), $this->userManager);
+		$this->cloudIdManager = new CloudIdManager(
+			$this->contactsManager,
+			$this->createMock(IURLGenerator::class),
+			$this->userManager,
+			$this->createMock(ICacheFactory::class),
+			$this->createMock(EventDispatcherInterface::class)
```
Comment (CarlSchwan) [inline apps/federatedfilesharing/tests/FederatedShareProviderTest.php:126]:
```suggestion
			$this->createMock(IEventDispatche::class)
```

File: apps/federatedfilesharing/tests/FederatedShareProviderTest.php [hunk 9]
```diff
@@ -51,6 +52,7 @@
 use OCP\Share\IManager;
 use OCP\Share\IShare;
 use PHPUnit\Framework\MockObject\MockObject;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
```
Comment (CarlSchwan) [inline apps/federatedfilesharing/tests/FederatedShareProviderTest.php:55]:
```suggestion
 use OCP\EventDispatcher\IEventDispatcher;
```

File: apps/federatedfilesharing/tests/Controller/MountPublicLinkControllerTest.php [hunk 10]
```diff
@@ -46,6 +47,7 @@
 use OCP\IUserSession;
 use OCP\Share\IManager;
 use OCP\Share\IShare;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
```
Comment (CarlSchwan) [inline apps/federatedfilesharing/tests/Controller/MountPublicLinkControllerTest.php:50]:
```suggestion
 use OCP\EventDispatcher\IEventDispatcher;
```

File: apps/federatedfilesharing/tests/Controller/MountPublicLinkControllerTest.php [hunk 11]
```diff
@@ -107,7 +109,13 @@ protected function setUp(): void {
 		$this->userSession = $this->getMockBuilder(IUserSession::class)->disableOriginalConstructor()->getMock();
 		$this->clientService = $this->getMockBuilder('OCP\Http\Client\IClientService')->disableOriginalConstructor()->getMock();
 		$this->contactsManager = $this->createMock(IContactsManager::class);
-		$this->cloudIdManager = new CloudIdManager($this->contactsManager, $this->createMock(IURLGenerator::class), $this->userManager);
+		$this->cloudIdManager = new CloudIdManager(
+			$this->contactsManager,
+			$this->createMock(IURLGenerator::class),
+			$this->userManager,
+			$this->createMock(ICacheFactory::class),
+			$this->createMock(EventDispatcherInterface::class)
```
Comment (CarlSchwan) [inline apps/federatedfilesharing/tests/Controller/MountPublicLinkControllerTest.php:117]:
```suggestion
			$this->createMock(IEventDispatcher::class)
```

File: lib/private/Server.php [hunk 12]
```diff
@@ -1361,7 +1361,13 @@ public function __construct($webRoot, \OC\Config $config) {
 		});
 
 		$this->registerService(ICloudIdManager::class, function (ContainerInterface $c) {
-			return new CloudIdManager($c->get(\OCP\Contacts\IManager::class), $c->get(IURLGenerator::class), $c->get(IUserManager::class));
+			return new CloudIdManager(
+				$c->get(\OCP\Contacts\IManager::class),
+				$c->get(IURLGenerator::class),
+				$c->get(IUserManager::class),
+				$c->get(ICacheFactory::class),
+				$c->get(EventDispatcherInterface::class),
```
Comment (CarlSchwan) [inline lib/private/Server.php:1369]:
```suggestion
				$c->get(IEventDispatcher::class),
```

File: tests/lib/Collaboration/Collaborators/MailPluginTest.php [hunk 13]
```diff
@@ -38,6 +39,7 @@
 use OCP\IUserSession;
 use OCP\Share\IShare;
 use OCP\Mail\IMailer;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
```
Comment (CarlSchwan) [inline tests/lib/Collaboration/Collaborators/MailPluginTest.php:42]:
```suggestion
 use OCP\EventDispatcher\IEventDispatcher;
```

File: tests/lib/Collaboration/Collaborators/MailPluginTest.php [hunk 14]
```diff
@@ -77,7 +79,13 @@ protected function setUp(): void {
 		$this->knownUserService = $this->createMock(KnownUserService::class);
 		$this->userSession = $this->createMock(IUserSession::class);
 		$this->mailer = $this->createMock(IMailer::class);
-		$this->cloudIdManager = new CloudIdManager($this->contactsManager, $this->createMock(IURLGenerator::class), $this->createMock(IUserManager::class));
+		$this->cloudIdManager = new CloudIdManager(
+			$this->contactsManager,
+			$this->createMock(IURLGenerator::class),
+			$this->createMock(IUserManager::class),
+			$this->createMock(ICacheFactory::class),
+			$this->createMock(EventDispatcherInterface::class)
```
Comment (CarlSchwan) [inline tests/lib/Collaboration/Collaborators/MailPluginTest.php:87]:
```suggestion
			$this->createMock(IEventDispatcher::class)
```

File: tests/lib/Collaboration/Collaborators/RemotePluginTest.php [hunk 15]
```diff
@@ -29,12 +29,14 @@
...
 use OCP\Contacts\IManager;
 use OCP\Federation\ICloudIdManager;
+use OCP\ICacheFactory;
 use OCP\IConfig;
 use OCP\IURLGenerator;
 use OCP\IUser;
 use OCP\IUserManager;
 use OCP\IUserSession;
 use OCP\Share\IShare;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
```
Comment (CarlSchwan) [inline tests/lib/Collaboration/Collaborators/RemotePluginTest.php:39]:
```suggestion
 use OCP\EventDispatcher\IEventDispatcher;
```

File: tests/lib/Collaboration/Collaborators/RemotePluginTest.php [hunk 16]
```diff
@@ -63,7 +65,13 @@ protected function setUp(): void {
 		$this->userManager = $this->createMock(IUserManager::class);
 		$this->config = $this->createMock(IConfig::class);
 		$this->contactsManager = $this->createMock(IManager::class);
-		$this->cloudIdManager = new CloudIdManager($this->contactsManager, $this->createMock(IURLGenerator::class), $this->createMock(IUserManager::class));
+		$this->cloudIdManager = new CloudIdManager(
+			$this->contactsManager,
+			$this->createMock(IURLGenerator::class),
+			$this->createMock(IUserManager::class),
+			$this->createMock(ICacheFactory::class),
+			$this->createMock(EventDispatcherInterface::class)
```
Comment (CarlSchwan) [inline tests/lib/Collaboration/Collaborators/RemotePluginTest.php:73]:
```suggestion
			$this->createMock(IEventDispatcher::class)
```

File: tests/lib/Federation/CloudIdManagerTest.php [hunk 17]
```diff
@@ -22,9 +22,12 @@
 namespace Test\Federation;
 
 use OC\Federation\CloudIdManager;
+use OC\Memcache\ArrayCache;
 use OCP\Contacts\IManager;
+use OCP\ICacheFactory;
 use OCP\IURLGenerator;
 use OCP\IUserManager;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
```
Comment (CarlSchwan) [inline tests/lib/Federation/CloudIdManagerTest.php:30]:
```suggestion
 use OCP\EventDispatcher\IEventDispatcher;
```

File: tests/lib/Federation/CloudIdManagerTest.php [hunk 18]
```diff
@@ -45,7 +50,17 @@ protected function setUp(): void {
...
+		$this->cacheFactory = $this->createMock(ICacheFactory::class);
+		$this->cacheFactory->method('createLocal')
+			->willReturn(new ArrayCache(''));
+
+		$this->cloudIdManager = new CloudIdManager(
+			$this->contactsManager,
+			$this->urlGenerator,
+			$this->userManager,
+			$this->cacheFactory,
+			$this->createMock(EventDispatcherInterface::class)
```
Comment (CarlSchwan) [inline tests/lib/Federation/CloudIdManagerTest.php:62]:
```suggestion
			$this->createMock(IEventDispatcher::class)
```

**Review by CarlSchwan — APPROVED** [review]
Need to adjusts the test too ;)

**Pull request merged**